py: Use correct exception message for implicit conversion to str or b…

stinos opened the pull request:
…ytes

When terse error reporting is not active use the actual target type in
the exception message instead of always reporting conversion 'to str'
fails even if it happens to actually be 'to bytes'. See #2957.

Note in the 3 functions near the end of the file I used 'str' as target type no matter what but that seems reasonable.

Also maybe it's worth having a check_implicit_conversion to reduce some duplication? Like
```
void check_bad_implicit_conversion(mp_obj_t src, mp_obj_type_t *targetType) {
    mp_obj_type_t *srcType = mp_obj_get_type(src);
    if (srcType  != targetType) {
        bad_implicit_conversion(srcType, targetType);
    }
}
```
so all occurrences of
```
if (mp_obj_get_type(args[1]) != self_type) {
    bad_implicit_conversion(args[1], self_type);
}
```
can be replaced with just ```check_bad_implicit_conversion(args[1], self_type)```

pfalcon commented:
Oops, didn't think it would take so much red/green paint ;-). Did you try to measure code size growth?

> can be replaced with just check_bad_implicit_conversion(args[1], self_type)

That might be good idea to compensate back the growth, but it costs more cycles and potentially stack space...

dpgeorge commented:
An alternative is to always use the existing message, but just make sure that the types that are printed are always different.  So if the incoming argument is of type str then actually print "bytes".  That would mean the error message is "backwards" (says it can convert bytes to str when it actually means it can't convert str to bytes) but at least this is more user friendly than the current behaviour.

stinos commented:
@pfalcon Patch as-is costs 80bytes, one extra push for the second argument. Using a seperate check_bad_implicit_conversion function brings this down to 68bytes (but not on msvc as it gets inlined) and again an extra push when conversion actually fails and bad_implicit_conversion gets called.

@dpgeorge your idea is 48 bytes. And actually the error message isn't backwards for str<->bytes, it's only 'slightly' incorrect when trying to for example convert int to bytes, in which case it will complain about int to str conversion instead; not that big of a problem I'd say.

bad_implicit_conversion then has:
```
const qstr src_name = mp_obj_get_type(self_in)->name;
nlr_raise(mp_obj_new_exception_msg_varg(&mp_type_TypeError,
    "can't convert '%s' object to %s implicitly",
    qstr_str(src_name), qstr_str(src_name == MP_QSTR_str ? MP_QSTR_bytes : MP_QSTR_str)));
```

and exceptions are:
```
>>> 'b'.count(b'g')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: can't convert 'bytes' object to str implicitly

>>> b'b'.count('g')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: can't convert 'str' object to bytes implicitly

>>> 'b'.count(1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: can't convert 'int' object to str implicitly

>>> b'b'.count(1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: can't convert 'int' object to str implicitly
```

It's up to you guys, I'd go for the latter solution.

dpgeorge commented:
@stinos I was actually thinking of this:
```
        qstr src_name = mp_obj_get_type(self_in)->name;
        nlr_raise(mp_obj_new_exception_msg_varg(&mp_type_TypeError,
            "can't convert '%q' object to str implicitly",
            src_name == MP_QSTR_str ? MP_QSTR_bytes : src_name));
```
So it always prints str as the second type, and the first one is never str.  I like your solution better because it costs not much more and gives better error messages in some cases.

Note the use of "%q" above, you can do this to eliminate calls to qstr_str.

stinos committed:
py: Use better error message in bad implicit str/bytes conversion exception

Instead of always reporting some object cannot be implicitly be converted
to a 'str', even when it is a 'bytes' object, adjust the logic so that
when trying to convert str to bytes it is shown like that.
This will still report bad implicit conversion from e.g. 'int to bytes'
as 'int to str' but it will not result in the confusing
'can't convert 'str' object to str implicitly' anymore for calls like
b'somestring'.count('a').

stinos commented:
Ok I pushed the change, including use of %q

dpgeorge commented:
Thanks, merged in bf29fe2e138a30688cfb94ad3859f903f935ced1.
